File: app/src/main/java/com/iceteaviet/fastfoodfinder/activity/MapRoutingActivity.java
package com.iceteaviet.fastfoodfinder.activity;

import android.os.Bundle;
import android.support.design.widget.BottomSheetBehavior;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.LinearSnapHelper;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SnapHelper;
import android.support.v7.widget.Toolbar;
import android.util.DisplayMetrics;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.maps.android.PolyUtil;
import com.iceteaviet.fastfoodfinder.R;
import com.iceteaviet.fastfoodfinder.adapter.RoutingAdapter;
import com.iceteaviet.fastfoodfinder.helper.DividerItemDecoration;
import com.iceteaviet.fastfoodfinder.model.Routing.MapsDirection;
import com.iceteaviet.fastfoodfinder.model.Routing.Step;
import com.iceteaviet.fastfoodfinder.model.Store.Store;
import com.iceteaviet.fastfoodfinder.utils.DisplayUtils;
import com.iceteaviet.fastfoodfinder.utils.Keys;
import com.iceteaviet.fastfoodfinder.utils.MapUtils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MapRoutingActivity extends AppCompatActivity {

    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.tv_routing_time)
    TextView txtTravelTime;
    @BindView(R.id.tv_routing_distance)
    TextView txtTravelDistance;
    @BindView(R.id.tv_routing_overview)
    TextView txtTravelOverview;
    @BindView(R.id.rv_bottom_sheet)
    RecyclerView bottomRecyclerView;
    @BindView(R.id.rv_direction_instruction)
    RecyclerView topRecyclerView;
    @BindView(R.id.ll_bottom_sheet)
    LinearLayout mBottomSheetContainer;
    @BindView(R.id.btn_prev_instruction)
    ImageButton prevInstruction;
    @BindView(R.id.btn_next_instruction)
    ImageButton nextInstruction;
    @BindView(R.id.ll_routing_button_container)
    LinearLayout routingButtonContainer;
    RoutingAdapter.OnNavigationItemClickListener mListener;
    DividerItemDecoration divider;
    boolean isPreviewMode = false;
    private BottomSheetBehavior mBottomSheetBehavior;
    private GoogleMap mGoogleMap;
    private SupportMapFragment mMapFragment;
    private MapsDirection mMapsDirection;
    private Polyline mCurrDirection;
    private LatLng mCurrLocation;
    private Store mCurrStore;
    private RoutingAdapter mBottomRoutingAdapter;
    private RoutingAdapter mTopRoutingAdapter;
    private List<Step> mStepList;
    private List<LatLng> mGeoPointList;
    private int currDirectionIndex = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_map_routing);

        ButterKnife.bind(this);

        mStepList = new ArrayList<>();
        mGeoPointList = new ArrayList<>();

        mBottomSheetBehavior = BottomSheetBehavior.from(mBottomSheetContainer);

        getExtrasBundle();

        setupToolbar(mToolbar);

        mStepList = mMapsDirection.getRouteList().get(0).getLegList().get(0).getStepList();
        mCurrLocation = mStepList.get(0).getStartMapCoordination().getLocation();
        mGeoPointList = PolyUtil.decode(mMapsDirection.getRouteList().get(0).getEncodedPolylineString());


        mBottomRoutingAdapter = new RoutingAdapter(mStepList, RoutingAdapter.TYPE_FULL);
        LinearLayoutManager layoutManager = new LinearLayoutManager(MapRoutingActivity.this);
        bottomRecyclerView.setLayoutManager(layoutManager);
        bottomRecyclerView.setAdapter(mBottomRoutingAdapter);

        mTopRoutingAdapter = new RoutingAdapter(mStepList, RoutingAdapter.TYPE_SHORT);
        final LinearLayoutManager topLayoutManager = new LinearLayoutManager(MapRoutingActivity.this, LinearLayoutManager.HORIZONTAL, false);
        topRecyclerView.setLayoutManager(topLayoutManager);
        topRecyclerView.setAdapter(mTopRoutingAdapter);
        SnapHelper snapHelper = new LinearSnapHelper();
        snapHelper.attachToRecyclerView(topRecyclerView);

        mListener = new RoutingAdapter.OnNavigationItemClickListener() {
            @Override
            public void onClick(int index) {
                showDirectionAt(index);
                currDirectionIndex = index;
                enterPreviewMode();
            }
        };

        prevInstruction.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                currDirectionIndex--;
                if (currDirectionIndex >= 0 && currDirectionIndex < mBottomRoutingAdapter.getItemCount()) {
                    topRecyclerView.smoothScrollToPosition(currDirectionIndex);
                    showDirectionAt(currDirectionIndex);
                } else {
                    currDirectionIndex = 0;
                }
            }
        });

        nextInstruction.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                currDirectionIndex++;
                if (currDirectionIndex >= 0 && currDirectionIndex < mBottomRoutingAdapter.getItemCount()) {
                    topRecyclerView.smoothScrollToPosition(currDirectionIndex);
                    showDirectionAt(currDirectionIndex);
                } else {
                    currDirectionIndex = mBottomRoutingAdapter.getItemCount() - 1;
                }
            }
        });


        setUpMapIfNeeded();
    }

    private void showDirectionAt(int currDirectionIndex) {
        mGoogleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(mBottomRoutingAdapter.getDirectionLocationAt(currDirectionIndex), 18));
    }

    private void getExtrasBundle() {
        Bundle extras = getIntent().getExtras();
        mMapsDirection = extras.getParcelable(Keys.KEY_ROUTE_LIST);
        mCurrStore = extras.getParcelable(Keys.KEY_DES_STORE);

        if (mMapsDirection == null || mCurrStore == null || mMapsDirection.getRouteList().size() <= 0) {
            Toast.makeText(MapRoutingActivity.this, "Failed to open Routing screen!", Toast.LENGTH_SHORT).show();
            finish();
        }
    }

    private void setupToolbar(Toolbar toolbar) {
        setSupportActionBar(toolbar);

        // add back arrow to mToolbar
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setTitle(mCurrStore.getTitle());
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // handle arrow click here
        if (item.getItemId() == android.R.id.home) {
            if (isPreviewMode) {
                exitPreviewMode();
            } else {
                finish(); // close this activity and return to preview activity (if there is any)

            }
        }

        return super.onOptionsItemSelected(item);
    }

    protected void setUpMapIfNeeded() {
        // Do a null check to confirm that we have not already instantiated the map.
        if (mMapFragment == null) {
            mMapFragment = ((SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map));
            // Check if we were successful in obtaining the map.
            if (mMapFragment != null) {
                mMapFragment.getMapAsync(new OnMapReadyCallback() {
                    @Override
                    public void onMapReady(GoogleMap map) {
                        loadMap(map);
                    }
                });
            }
        }
    }

    // The Map is verified. It is now safe to manipulate the map.
    protected void loadMap(GoogleMap googleMap) {
        if (googleMap != null) {
            // ... use map here
            mGoogleMap = googleMap;
            mGoogleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(mCurrLocation, 16));

            Marker currMarker = googleMap.addMarker(new MarkerOptions().position(mCurrLocation)
                    .title("Your location")
                    .snippet("Your current location, please follow the line")
                    .icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_map_bluedot)));

            if (mCurrStore != null && mCurrStore.getPosition() != null) {
                Marker storeMarker = googleMap.addMarker(new MarkerOptions().position(mCurrStore.getPosition())
                        .title(mCurrStore.getTitle())
                        .snippet(mCurrStore.getAddress())
                        .icon(BitmapDescriptorFactory.fromResource(MapUtils.getLogoDrawableId(mCurrStore.getType()))));
            }

            drawPolylines(mGeoPointList, mGoogleMap);


            txtTravelTime.setText(mMapsDirection.getRouteList().get(0).getLegList().get(0).getDuration());
            txtTravelDistance.setText(mMapsDirection.getRouteList().get(0).getLegList().get(0).getDistance());
            txtTravelOverview.setText("Via " + mMapsDirection.getRouteList().get(0).getSummary());


            mBottomRoutingAdapter.setOnNavigationItemClickListener(mListener);
            mTopRoutingAdapter.setOnNavigationItemClickListener(mListener);
        }
    }

    void drawPolylines(List<LatLng> geoPointList, GoogleMap googleMap) {
        //Add position to viewBounds
        LatLngBounds.Builder builder = new LatLngBounds.Builder();
        builder.include(mCurrLocation);

        PolylineOptions options = new PolylineOptions()
                .clickable(true)
                .color(ContextCompat.getColor(MapRoutingActivity.this, R.color.googleBlue))
                .width(12)
                .geodesic(true)
                .zIndex(5f);

        for (int i = 0; i < geoPointList.size(); i++) {
            LatLng geoPoint = geoPointList.get(i);
            options.add(geoPoint);
            builder.include(geoPoint);
        }

        if (mCurrDirection != null)
            mCurrDirection.remove();

        mCurrDirection = googleMap.addPolyline(options);

        //Build the viewbounds contain all markers
        LatLngBounds bounds = builder.build();

        zoomToShowAllMarker(bounds, googleMap);

    }

    private void zoomToShowAllMarker(LatLngBounds bounds, GoogleMap googleMap) {
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();

        int width = displayMetrics.widthPixels;
        int height = displayMetrics.heightPixels - DisplayUtils.convertDpToPx(displayMetrics, 160);
        int padding = 24; // offset from edges of the map in pixels
        CameraUpdate cu = CameraUpdateFactory.newLatLngBounds(bounds, width, height, padding);

        googleMap.animateCamera(cu);
    }

    private void enterPreviewMode() {
        routingButtonContainer.setVisibility(View.VISIBLE);
        topRecyclerView.setVisibility(View.VISIBLE);
        mBottomSheetBehavior.setHideable(true);
        mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);
        isPreviewMode = true;
    }

    private void exitPreviewMode() {
        routingButtonContainer.setVisibility(View.GONE);
        topRecyclerView.setVisibility(View.GONE);
        mBottomSheetBehavior.setHideable(false);
        mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
        isPreviewMode = false;
    }
}

File: app/src/main/java/com/iceteaviet/fastfoodfinder/ui/routing/MapRoutingActivity.kt
package com.iceteaviet.fastfoodfinder.ui.routing

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSnapHelper
import androidx.recyclerview.widget.RecyclerView
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.Polyline
import com.google.android.gms.maps.model.PolylineOptions
import com.google.android.material.bottomsheet.BottomSheetBehavior
import dagger.hilt.android.AndroidEntryPoint
import com.iceteaviet.fastfoodfinder.R
import com.iceteaviet.fastfoodfinder.data.remote.routing.model.MapsDirection
import com.iceteaviet.fastfoodfinder.data.remote.store.model.Store
import com.iceteaviet.fastfoodfinder.databinding.ActivityMapRoutingBinding
import com.iceteaviet.fastfoodfinder.ui.base.BaseActivity
import com.iceteaviet.fastfoodfinder.ui.custom.snaphelper.OnSnapListener
import com.iceteaviet.fastfoodfinder.ui.custom.snaphelper.OnSnapPositionChangeListener
import com.iceteaviet.fastfoodfinder.utils.Constant.DEFAULT_ZOOM_LEVEL
import com.iceteaviet.fastfoodfinder.utils.Constant.DETAILED_ZOOM_LEVEL
import com.iceteaviet.fastfoodfinder.utils.convertDpToPx
import com.iceteaviet.fastfoodfinder.utils.extension.attachSnapHelperToListener
import com.iceteaviet.fastfoodfinder.utils.ui.getStoreLogoDrawableRes
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

@AndroidEntryPoint
class MapRoutingActivity : BaseActivity(), View.OnClickListener {

    private val viewModel: MapRoutingViewModel by viewModels()

    private lateinit var binding: ActivityMapRoutingBinding

    private lateinit var txtTravelTime: TextView
    private lateinit var txtTravelDistance: TextView
    private lateinit var txtTravelOverview: TextView
    private lateinit var bottomRecyclerView: RecyclerView
    private lateinit var topRecyclerView: RecyclerView
    private lateinit var bottomSheetContainer: LinearLayout
    private lateinit var prevInstruction: ImageButton
    private lateinit var nextInstruction: ImageButton
    private lateinit var routingButtonContainer: LinearLayout

    private val snapHelper = LinearSnapHelper()
    private var bottomSheetBehavior: BottomSheetBehavior<*>? = null
    private var googleMap: GoogleMap? = null
    private var mapFragment: SupportMapFragment? = null
    private var currDirection: Polyline? = null
    private lateinit var bottomRoutingAdapter: RoutingAdapter
    private lateinit var topRoutingAdapter: RoutingAdapter

    override val layoutId: Int
        get() = R.layout.activity_map_routing

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMapRoutingBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupUI()
        setUpMapIfNeeded()
        setupEventListeners()

        intent.extras?.let {
            viewModel.handleExtras(
                it.getParcelable<MapsDirection>(KEY_ROUTE_LIST), 
                it.getParcelable<Store>(KEY_DES_STORE)
            )
        }
        
        setupObservers()
    }

    private fun setupObservers() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collectLatest { state ->
                    supportActionBar?.title = state.title
                    
                    bottomRoutingAdapter.setStepList(state.stepList)
                    topRoutingAdapter.setStepList(state.stepList)

                    if (state.inPreviewMode) {
                        enterPreviewMode()
                    } else {
                        exitPreviewMode()
                    }

                    txtTravelTime.text = state.durationText
                    txtTravelDistance.text = state.distanceText
                    txtTravelOverview.text = state.summaryText

                    when (val event = state.event) {
                        is MapRoutingEvent.Idle -> {}
                        is MapRoutingEvent.Exit -> {
                            finish()
                            viewModel.markEventConsumed()
                        }
                        is MapRoutingEvent.ShowGetDirectionFailedMessage -> {
                            Toast.makeText(this@MapRoutingActivity, R.string.get_map_direction_failed, Toast.LENGTH_SHORT).show()
                            finish() // Exit automatically on fail
                            viewModel.markEventConsumed()
                        }
                        is MapRoutingEvent.ShowGeneralErrorMessage -> {
                            Toast.makeText(this@MapRoutingActivity, R.string.error_general_error_code, Toast.LENGTH_LONG).show()
                            viewModel.markEventConsumed()
                        }
                        is MapRoutingEvent.AnimateMapCamera -> {
                            animateMapCamera(event.location, event.zoomToDetail)
                            viewModel.markEventConsumed()
                        }
                        is MapRoutingEvent.AddMapMarker -> {
                            addMapMarker(event.location, event.title, event.description, event.icon)
                            viewModel.markEventConsumed()
                        }
                        is MapRoutingEvent.DrawRoutingPath -> {
                            drawRoutingPath(event.currLocation, event.routingGeoPoint)
                            viewModel.markEventConsumed()
                        }
                        is MapRoutingEvent.ScrollTopBannerToPosition -> {
                            scrollTopBannerToPosition(event.index)
                            // manually animate map here to fix conflation limits:
                            if(state.stepList.isNotEmpty()) {
                                animateMapCamera(state.stepList[event.index].endMapCoordination.location, true)
                            }
                            viewModel.markEventConsumed()
                        }
                    }
                }
            }
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_prev_instruction -> viewModel.onPrevInstructionClick()
            R.id.btn_next_instruction -> viewModel.onNextInstructionClick()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            viewModel.onBackArrowButtonPress()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun enterPreviewMode() {
        routingButtonContainer.visibility = View.VISIBLE
        topRecyclerView.visibility = View.VISIBLE
        bottomSheetBehavior?.isHideable = true
        bottomSheetBehavior?.state = BottomSheetBehavior.STATE_HIDDEN
    }

    private fun exitPreviewMode() {
        routingButtonContainer.visibility = View.GONE
        topRecyclerView.visibility = View.GONE
        bottomSheetBehavior?.isHideable = false
        bottomSheetBehavior?.state = BottomSheetBehavior.STATE_COLLAPSED
    }

    private fun scrollTopBannerToPosition(directionIndex: Int) {
        topRecyclerView.smoothScrollToPosition(directionIndex)
    }

    private fun animateMapCamera(location: LatLng, zoomToDetail: Boolean) {
        val zoomLevel = if (zoomToDetail) DETAILED_ZOOM_LEVEL else DEFAULT_ZOOM_LEVEL
        googleMap?.animateCamera(CameraUpdateFactory.newLatLngZoom(location, zoomLevel))
    }

    private fun addMapMarker(location: LatLng, title: String, description: String, icon: Int) {
        googleMap?.addMarker(MarkerOptions().position(location)
            .title(title)
            .snippet(description)
            .icon(BitmapDescriptorFactory.fromResource(icon)))
    }

    private fun drawRoutingPath(currLocation: LatLng?, routingGeoPoint: List<LatLng>) {
        googleMap?.let {
            val builder = LatLngBounds.Builder()
            if (currLocation != null) builder.include(currLocation)

            val options = PolylineOptions()
                .clickable(true)
                .color(ContextCompat.getColor(this, R.color.googleBlue))
                .width(12f)
                .geodesic(true)
                .zIndex(5f)

            for (i in routingGeoPoint.indices) {
                val geoPoint = routingGeoPoint[i]
                options.add(geoPoint)
                builder.include(geoPoint)
            }

            currDirection?.remove()
            currDirection = it.addPolyline(options)

            val bounds = builder.build()
            val padding = 48
            zoomToShowAllMarker(bounds, it, padding)
        }
    }

    private fun setupUI() {
        findViews()
        bottomSheetBehavior = BottomSheetBehavior.from(bottomSheetContainer)

        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)

        bottomRoutingAdapter = RoutingAdapter(RoutingAdapter.TYPE_FULL)
        val layoutManager = LinearLayoutManager(this)
        bottomRecyclerView.layoutManager = layoutManager
        bottomRecyclerView.adapter = bottomRoutingAdapter

        topRoutingAdapter = RoutingAdapter(RoutingAdapter.TYPE_SHORT)
        val topLayoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        topRecyclerView.layoutManager = topLayoutManager
        topRecyclerView.adapter = topRoutingAdapter
        snapHelper.attachToRecyclerView(topRecyclerView)
    }

    private fun findViews() {
        txtTravelTime = binding.tvRoutingTime
        txtTravelDistance = binding.tvRoutingDistance
        txtTravelOverview = binding.tvRoutingOverview
        bottomRecyclerView = binding.rvBottomSheet
        topRecyclerView = binding.rvDirectionInstruction
        bottomSheetContainer = binding.llBottomSheet
        prevInstruction = binding.btnPrevInstruction
        nextInstruction = binding.btnNextInstruction
        routingButtonContainer = binding.llRoutingButtonContainer
    }

    private fun setupEventListeners() {
        prevInstruction.setOnClickListener(this)
        nextInstruction.setOnClickListener(this)

        val listener = object : RoutingAdapter.OnNavigationRowClickListener {
            override fun onClick(index: Int) {
                viewModel.onNavigationRowClick(index)
            }
        }
        bottomRoutingAdapter.setOnNavigationItemClickListener(listener)
        topRoutingAdapter.setOnNavigationItemClickListener(listener)

        topRecyclerView.attachSnapHelperToListener(snapHelper, object : OnSnapPositionChangeListener {
            override fun onSnapPositionChange(position: Int) {
                viewModel.onTopRoutingBannerPositionChange(position)
            }
        }, OnSnapListener.Behavior.NOTIFY_ON_SCROLL_STATE_IDLE_BY_DRAGGING)
    }

    private fun setUpMapIfNeeded() {
        if (mapFragment === null) {
            mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment

            mapFragment?.getMapAsync { map ->
                googleMap = map
                viewModel.onGetMapAsync()
                
                // Directly pull map init data here to render map correctly bypassing event limits
                val mapData = viewModel.getInitMapData()
                if (mapData != null) {
                    addMapMarker(mapData.store.getPosition(), mapData.store.title, mapData.store.address, getStoreLogoDrawableRes(mapData.store.type))
                    mapData.currLocation?.let { loc ->
                        animateMapCamera(loc, false)
                        addMapMarker(loc, "Your location", "Your current location, please follow the line", R.drawable.ic_map_bluedot)
                    }
                    drawRoutingPath(mapData.currLocation, mapData.geoPointList)
                }
            }
        }
    }

    private fun zoomToShowAllMarker(bounds: LatLngBounds, googleMap: GoogleMap, padding: Int) {
        val displayMetrics = resources.displayMetrics

        val width = displayMetrics.widthPixels
        val height = displayMetrics.heightPixels - convertDpToPx(160f)
        val cu = CameraUpdateFactory.newLatLngBounds(bounds, width, height.toInt(), padding)

        googleMap.animateCamera(cu)
    }

    companion object {
        const val KEY_ROUTE_LIST = "route_list"
        const val KEY_DES_STORE = "des_store"
    }
}